value/src/main/java/com/google/auto/value/AutoValueExtension.java

```diff
@@ -0,0 +1,33 @@
+package com.google.auto.value;
+
+import javax.annotation.processing.ProcessingEnvironment;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.TypeElement;
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * Created by rharter on 5/1/15.
+ */
+public interface AutoValueExtension {
+
+  public interface Context {
+    ProcessingEnvironment processingEnvironment();
+    String packageName();
+    TypeElement autoValueClass();
+    Map<String, ExecutableElement> properties();
+  }
+
+  public interface GeneratedClass {
+    String className();
+    String source();
+    Collection<String> additionalImports();
+    Collection<ExecutableElement> consumedProperties();
+    Collection<String> additionalInterfaces();
+    Collection<String> additionalCode();
```

Comment:
When I implemented the subclassing version of this, the ordering was actually different.  I treated the extension classes as the intermediaries, so AutoValue_Foo is the AutoValue implementation, which extends $AutoValue_Foo which is the AutoParcel implementation, which extends Foo.

This is important, particularly if you have more than one extension that requires things like static factory methods or static properties.  Even if you have multiple plugins and only one requires static properties, we'd need to devise a prioritization mechanism for the extensions.

That's why I implemented this with both the ability to add intermediate classes, and the ability to add code to the final concrete AutoValue class.